File: c++Poker/playTable.cpp
#include <iostream>
#include <list>
//#include "Card.cpp"
#include "Player.h"
using namespace std;
#include "Deck.cpp"
#include <fstream>
#include <sstream>

static void ShowWinners(Player players[]);
static void tieBreak(vector<Player> &v);

int main(int argc, char* argv[])
{

    
    if(argv[1]!=NULL)
    {
        //cout<<argv[1];
        ifstream file(argv[1]);
        Player players[6];

        if (file) 
        {
            string line;
            int a=0;
            while (getline(file, line)) {
                
                stringstream ss(line);
                Card cards[5];
                string card;
                int i=0;
                while (getline(ss, card, ',' )) {
                    
                    card.erase(remove_if(card.begin(), card.end(), ::isspace), card.end());
                    cards[i]=card;
                    i++;
                }

                Player player(cards);
                //player.setHand(cards);
                players[a]=player;
                a++;
            }

        }
        cout<<"*** Using Test Deck ***\n\n";
        cout<<"*** File:  "<<argv[1]<<endl;
        for(Player &p : players)
        {
            p.showHand();
            p.rankHand();
        }
        cout<<"\n---Winning Hand Order---\n";
        ShowWinners(players);


    }
    else
    {        
        Deck d;
        d.shuffle();
        cout<<"*** USING RANDOMIZED DECK OF CARDS ***\n\n";
        cout<<"*** Shuffled 52 Card Deck: \n";
        d.showDeck();

        Player table[6];
        for (int i=0; i<6; i++)
        {
            Card temp[5];
            for(int a=0; a<5; a++)
            {
                temp[a]=d.popCard();
            }
            table[i]= Player(temp);
        }

        cout<<"*** Here are the six hands ... \n";

        for (Player &p : table)
        {
            p.showHand();
            p.rankHand();
        }
      
        cout<<"\n*** Here is what remains in the deck: \n";
        d.showDeck();

        cout<<'\n'<<endl;

        cout<<"---WINNING HAND ORDER---\n";
        ShowWinners(table);

        
    }


}


static void ShowWinners(Player players[6])
{
    vector <vector<Player>> rankTiers;
    
    for(int i=0; i<10; i++)
    {
        rankTiers.push_back(vector<Player>());
    }

    for(int i=0; i<6; i++)
    {
        rankTiers[players[i].numRank].push_back(players[i]);
    }

    for(vector<Player> &v : rankTiers)
    {
        if(v.size()>1)
        {
            tieBreak(v);
        }
    }
    for(vector<Player> v : rankTiers)
    {
        if(!v.empty())
        {
            for(Player p : v)
            {
                p.showHand();
            }
        }
    }



}


static void tieBreak(vector<Player> &v) //assuming there are at least 2 items in the vector we are changing the array so pass by reference
{
    switch(v[0].numRank)
    {
        case 0:
            sort(v.begin(), v.end(), Player::compareBySuit);
            reverse(v.begin(), v.end());
            break;
        case 1:
            sort(v.begin(), v.end(), Player::compareByHighCard);//
            reverse(v.begin(), v.end());
            break;
        case 2:
            sort(v.begin(), v.end(), Player::compareByCard);//
            reverse(v.begin(), v.end());
            break;
        case 3:
            sort(v.begin(), v.end(), Player::compareByCard);//
            reverse(v.begin(), v.end());
            break;
        case 4:
            sort(v.begin(), v.end(), Player::compareBySuit);//
            reverse(v.begin(), v.end());
            break;
        case 5:
            sort(v.begin(), v.end(), Player::compareByHighCard);//
            reverse(v.begin(), v.end());
            break;
        case 6:
            sort(v.begin(), v.end(), Player::compareByCard);//
            reverse(v.begin(), v.end());
            break;
        case 7:
            sort(v.begin(), v.end(), Player::compareByCard);//Still need to sort out specific ties
            reverse(v.begin(), v.end());    
            break;
        case 8:
            for(Player &p : v)
            {   
                if(p.hRank[0].Face==1)
                {
                    p.hRank[0].Face=14;
                    //sort(p.hand.begin(), p.hand.end());
                }
            }
            sort(v.begin(), v.end(), Player::compareByCard);//
            reverse(v.begin(), v.end());
            break;
        case 9:
            for(Player &p : v)
            {   
                if(p.hand[0].Face==1)
                {
                    p.hand[0].Face=14;
                    //p.rankHand();
                    sort(p.hand.begin(), p.hand.end());
                }
            }
            sort(v.begin(), v.end(), Player::compareByHighCard);
            reverse(v.begin(), v.end());
            break;
    }

}

File: fortranPoker/playTable.f90
program main
    use class_Card
    use class_Deck
    use class_Player
    use mainFunctions
    implicit none
    character(100) :: filename
    character(88) :: card_str
    character(80) :: txtline
    type(Deck) :: deck1
    type(Player), dimension(6) :: table
    character(:), allocatable :: line, outline, word, duplicates, usedCards, temporaryArray
    character(30) :: temp
    
    integer :: index
    integer :: i, a, ios, unit
    type(Card) :: tempCard

    do index=1, size(table)
        call table(index)%CreatePlayer()
    end do

     ! Check for the command line argument
    if (command_argument_count() == 1) then
        call get_command_argument(1, filename)
        unit =1
        open(unit, file=filename, status="old")
        print*, "*** USING TEST DECK ***"

        print*

        print*,"*** File: ", filename
        print *
        print *,"*** Here are the six hands..."

        ! Loop through the lines in the CSV file
        do i = 1, 6
            ! Read a line containing the player's hand
            read(unit, '(A)', iostat=ios) txtline
            !print *, txtline
            if (ios /= 0) then
                print *, "Error reading line ", i
                exit
            end if
        
        ! Initialize outline to be same string as line, it will get overwritten in 
        ! the subroutine, but we need it for loop control
        outline = txtline
            index=1
            do while (len(outline) .ne. 0)
                call get_next_token( txtline, outline, word)
                !write(*,'(A)') trim(adjustl(word))
                temp=trim(word)
                temp=sweep_blanks(temp)
                !print*, temp
                txtline = outline
               

                call tempCard%makeCardfromString(trim(temp))
                table(i)%hand(index)=tempCard
                index=index+1
            end do
        end do
        do index=1, size(table)
            call table(index)%ShowHand()
            call table(index)%RankHand()
        end do

        print *, "---WINNING HAND ORDER---"
        call ShowWinners(table)

    ! Close the CSV file
    close(unit)
    else

        print *, "*** USING RANDOMIZED DECK OF CARDS ***"
        print *
        print *, "Shuffled 52 card deck:"
        call deck1%InitializeDeck()
        call deck1%ShuffleDeck()
        call deck1%ShowDeck()

       
        print *, "*** Here are the six hands..."

        do index=1,5
            do a=1,size(table)
                call deck1%PopCard(tempCard)
                table(a)%hand(index)=tempCard   
            end do
        end do

        do index=1, size(table)
            call table(index)%ShowHand()
            call table(index)%RankHand()
        end do

        print*
        print *, "*** Here is what remains in the deck..."

        call deck1%ShowDeck()

        print*
        print*
        print*, "--- WINNING HAND ORDER ---"
        
        call ShowWinners(table)


    end if

    contains
    character(30) function sweep_blanks(in_str)
        character(*), intent(in) :: in_str
        character(30) :: out_str
        character :: ch
        integer :: j

        out_str = " "
        do j=1, len_trim(in_str)
        ! get j-th char
        ch = in_str(j:j)
        if (ch .ne. " ") then
            out_str = trim(out_str) // ch
        endif
        sweep_blanks = out_str 
        end do
    end function sweep_blanks

end program main
